### DISCUSSION TITLE
Crop modes not working in YouTube default view

### DISCUSSION THREAD

**Discussion by markusd1984** [tamius-han/ultrawidify#150] [General]
Crop modes only seem to work in YT theatre and full screen view,

while stretch modes and video alignment work in default YT view just fine.

Thus I take it's either a bug or perhaps not yet implemented?

**Comment by tamius-han** [comment]
Not quite a bug. Things work properly/by design. 

Crop mode won't blindly crop away the letterbox — it will only crop as much as needed to fully display a video of given aspect ratio on screen.

In default youtube player, a 16:9 (native) video will be displayed in a 16:9 box. Since a 21:9 crop of that 16:9 video already occupies the biggest possible size that still fits within the 16:9 player box, no cropping occurs.

(Note that manual zooming is not subjected to such limitations)

**Comment by markusd1984** [comment]
I take it doesn't take letterboxed videos into considerations, example https://www.youtube.com/watch?v=Fua7w8ZJEWc
I was just surprised that it allowed to crop to 21:9 during theatre view but not default.


Since often one starts of from the default view, when one initially tries out the shortcuts D/S/W/E... without seeing any changes taking effect, one may think the extension isn't working. Hence I thought maybe it's only working in fulls screen (to have the most benefit of the cropping) but then was surprised to see it does actually work in theatre mode. 😄

**Comment by tamius-han** [comment]
> I take it doesn't take letterboxed videos into considerations

It does. However, you cannot crop away the letterbox without also cropping away the video, therefore cropping in such scenarios is super haram.

**Comment by markusd1984** [comment]
> > I take it doesn't take letterboxed videos into considerations
> 
> It does. However, you cannot crop away the letterbox without also cropping away the video, therefore cropping in such scenarios is super haram.

Not sure I'm following or perhaps there is some misunderstanding.

How is a 21:9 crop in a 16:9 video any different in the default YT view compared to the theatre view? Why wouldn't that have the same effect?

In other words why is this working in theatre mode and not in the default view?

**Comment by tamius-han** [comment]
The red rectangle is the video. The blue rectangle is the player. 

![slika](https://user-images.githubusercontent.com/12505802/114629999-e7657180-9cb9-11eb-8f11-0e5dbf0b2465.png)



Here's some ground rules:
1. we never ever _ever_ crop away any actual video, just the black bars
2. red rectangle must _always_ fit inside the blue rectangle
3. red rectangle must always be exactly as wide or exactly as tall as the blue rectangle
4. if rules 1-3 are violated, the red rectangle needs to be resized until rules 1-3 are no longer violated
5. ~~zoom/unzoom on z/u is exempt from the rules~~

Sometimes, the red rectangle is wrong. Which is where the crop mode comes in and says: "let's make this red rectangle into a 21:9 rectangle." Scissors are applied. Our red rectangle now looks like this.

![slika](https://user-images.githubusercontent.com/12505802/114631112-4cba6200-9cbc-11eb-8174-60f897bc1abb.png)




Now remember the ground rules?

In theater mode, we are now in violation of the rule \#3: red rectangle is not as tall as the blue rectangle, but it's also not as wide as the blue rectangle. Therefore, we need to zoom.
![slika](https://user-images.githubusercontent.com/12505802/114631319-a327a080-9cbc-11eb-997e-2dc15956d76c.png)


Meanwhile, on normal player, there's no horizontal space for video to grow. Video is already exactly as wide as the blue rectangle. The red rectangle — the video — cannot grow, because that would constitute a violation of rules 1 and 2:
![slika](https://user-images.githubusercontent.com/12505802/114631554-129d9000-9cbd-11eb-98eb-4554f1ab1490.png)

**Comment by markusd1984** [comment]
wow thanks for the great extensive explanation and pictures 👍 😄 

> Meanwhile, on normal player, there's no horizontal space for video to grow. Video is already exactly as wide as the blue rectangle. The red rectangle — the video — cannot grow, because that would constitute a violation of rules 1 and 2:
> ![slika](https://user-images.githubusercontent.com/12505802/114631554-129d9000-9cbd-11eb-98eb-4554f1ab1490.png)

I understand that in default view here scissors/crop is applied (red rectangle), does that mean the black part in the video is actually cropped away? (since you say no in rule number 1). 

I take the player height only changes according to the video source... if the video here is actually cropped than why can't it also shrink in height? Unless the player still sees the 16x9 video format.

**Comment by tamius-han** [comment]
> does that mean the black part in the video is actually cropped away? (since you say no in rule number 1).

No, it's not actually cropped away. Extension just pretends it is, because I've settled on using `transform: scale(x)` to scale the video, which is the easiest way to go (as `trasnform: scale(x,y)` provides you with the ability to scale width independently of height — something that is necessary for hybrid stretch).

In theory, `object-fit` exists, but doesn't offer the flexibility in scaling that I need.

**However** — there's also a lil bit of column B, because extension used to work by adjusting video width and height in the past (this was a thing before I switched to using `transform: scale()`, and I explicitly recall that changing the size of `<video>` didn't change the size of the player. Turns out that despite some evidence to the contrary, my memory still sometimes serves me right.

I went ahead and disabled the extension, using scaling video by setting width and height manually (side note: default height was 720px):

![image](https://user-images.githubusercontent.com/12505802/114937453-1c023600-9e3e-11eb-9598-94e68b66a625.png)

but the player remains the same, since youtube seemingly only calculates the player size either a) only when you switch between fullscreen/theater/normal or b) calculates player size using video.videoWidth and video.videoHeight.

![image](https://user-images.githubusercontent.com/12505802/114937754-7dc2a000-9e3e-11eb-8e5e-4ea18591b385.png)
